### PR TITLE
Create base types for tree nodes to use with Tree data structures

### PR DESCRIPTION
**Pull request by zsr2531**

### CODE REVIEW

File: src/Core/Echo.ControlFlow/Analysis/Domination/DominatorTreeNode.cs [hunk 1]
```diff
@@ -37,31 +27,14 @@ public INode OriginalNode
...
-
         /// <summary>
         /// Gets a collection of children representing all nodes that were dominated by the original node, as well as an
         /// immediate successor of the original node.
         /// </summary>
         /// <returns>The children, represented by the dominator tree nodes.</returns>
         public IEnumerable<DominatorTreeNode> GetDirectChildren()
         {
-            return Children.Where(c => c.OriginalNode.HasPredecessor(OriginalNode));
+            return Children.Where(c => c.Parent.HasPredecessor(OriginalNode)).Cast<DominatorTreeNode>();
```
Comment (Washi1337) [inline src/Core/Echo.ControlFlow/Analysis/Domination/DominatorTreeNode.cs:37]: Linq method chains on separate lines.

File: src/Core/Echo.ControlFlow/Analysis/Domination/DominatorTreeNode.cs [hunk 2]
```diff
@@ -71,7 +44,7 @@ public IEnumerable<DominatorTreeNode> GetDirectChildren()
         /// <returns>The children, represented by the dominator tree nodes.</returns>
         public IEnumerable<DominatorTreeNode> GetIndirectChildren()
         {
-            return Children.Where(c => !c.OriginalNode.HasPredecessor(OriginalNode));
+            return Children.Where(c => !c.Parent.HasPredecessor(OriginalNode)).Cast<DominatorTreeNode>();
```
Comment (Washi1337) [inline src/Core/Echo.ControlFlow/Analysis/Domination/DominatorTreeNode.cs:47]: Linq method chains on separate lines.

File: src/Core/Echo.ControlFlow/Analysis/Domination/DominatorTreeNode.cs [hunk 3]
```diff
@@ -92,36 +65,8 @@ public IEnumerable<DominatorTreeNode> GetDominatedNodes()
                 yield return current;
 
                 foreach (var child in current.Children)
-                    stack.Push(child);
+                    stack.Push(child as DominatorTreeNode);
```
Comment (Washi1337) [inline src/Core/Echo.ControlFlow/Analysis/Domination/DominatorTreeNode.cs:68]: Explicit cast instead of `as` cast.

**Comment by zsr2531** [comment]
@Washi1337 Addressed feedback.

**Pull request merged**